File: Java/345. Reverse Vowels of a String.java
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

class Solution {
    private static final Set<Character> vowels = new HashSet<>(Arrays.asList('a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'));

    public String reverseVowels(String s) {
        char[] chars = s.toCharArray();
        int left = 0;
        int right = chars.length - 1;

        while (left < right) {
            if (!resolver(chars[left])) {
                left++;
                continue;
            }

            if (!resolver(chars[right])) {
                right--;
                continue;
            }

            char newRight = chars[left];
            chars[left] = chars[right];
            chars[right] = newRight;

            left++;
            right--;
        }

        return new String(chars);
    }

    private boolean resolver(char c) {
        return vowels.contains(Character.toLowerCase((c)));
    }
}

File: CSharp/345. Reverse Vowels of a String.cs
public class Solution
{
    private static readonly HashSet<char> vowels = new() { 'a', 'e', 'i', 'o', 'u' };

    public string ReverseVowels(string s)
    {
        // 1. stringi char array'e çevir (string immutable)
        char[] chars = s.ToCharArray();
        
        // 2. left ve right pointer oluştur
        int left = 0;
        int right = s.Length - 1;

        // 3. sesli harfleri tanımlayan bir helper fonksiyon yaz
        // Resolver()

        // 4. while (left < right) döngüsünde:
        while (left < right)
        {
            //      - left ilerle, sesli değilse atla
            if (!Resolver(chars[left]))
            {
                left++;
                continue;
            }

            //      - right geri git, sesli değilse atla
            if (!Resolver(chars[right]))
            {
                right--;
                continue;
            }

            //      - ikisi de sesliyse swap
            char newRight = chars[left];
            chars[left] = chars[right];
            chars[right] = newRight;

            //      - ikisini ilerlet
            left++;
            right--;
        }

        // 5. yeni string döndür
        return new string(chars);
    }

    // Helper fonksiyon
    private bool Resolver(char c) => vowels.Contains(char.ToLower(c));
}